File: src/go/download-scotus-case-audio.go
package main

import (
    "bufio"
    "encoding/csv"
    "fmt"
    "io"
    "os"
    "os/exec"
    "path/filepath"
    "strings"
)

func main() {
    // Load the CSV file
    file, err := os.Open("./config/scotus.csv") // Replace with your actual file path
    if err != nil {
        fmt.Println("Error opening CSV file:", err)
        return
    }
    defer file.Close()

    reader := csv.NewReader(bufio.NewReader(file))

    // Base directory for the year
    baseDir := "../2023/"

    // Iterate over each row to generate and execute the curl commands
    for {
        record, err := reader.Read()
        if err == io.EOF {
            break
        }
        if err != nil {
            fmt.Println("Error reading CSV record:", err)
            continue
        }

        // Assuming 'Case ID' is in the first column
        caseID := strings.TrimSpace(record[0])

        // Full path for the case directory
        caseDir := filepath.Join(baseDir, caseID)

        // Create directory for the case if it doesn't exist
        audioDir := filepath.Join(caseDir, "audio")
        if err := os.MkdirAll(audioDir, os.ModePerm); err != nil {
            fmt.Println("Error creating directory:", err)
            continue
        }

        // Generate the curl command
        curlCommand := fmt.Sprintf("curl https://www.supremecourt.gov/media/audio/mp3files/%s.mp3 --output %s/audio/%s.mp3", caseID, caseDir, caseID)

        // Check if the file already exists
        filePath := filepath.Join(audioDir, fmt.Sprintf("%s.mp3", caseID))
        if _, err := os.Stat(filePath); os.IsNotExist(err) {
            // File doesn't exist, execute the curl command
            cmd := exec.Command("bash", "-c", curlCommand)
            if err := cmd.Run(); err != nil {
                fmt.Println("Error executing curl command:", err)
            }
        } else {
            // File already exists, skip downloading
            fmt.Println("Skipping download, file already exists:", filePath)
        }
    }
}

File: src/download-scotus-case-audio.py
import pandas as pd
import os
import subprocess
from datetime import datetime

def download_scotus_case_audio():
    """
    Downloads audio files for Supreme Court cases from the given CSV file.
    
    The function reads a CSV file containing case details and iterates over each row.
    For each case, it parses the argument date to create a directory by year, month, and case ID,
    then downloads the corresponding audio file into this directory using curl command.
    If the file already exists, it skips downloading.
    
    Args:
        None
        
    Returns:
        None
    """
    # Load the CSV file
    file_path = './config/scotus.csv' # Adjust path as needed
    scotus_df = pd.read_csv(file_path)

    # Get the current date
    current_date = datetime.now()

    # Iterate over each row to process the case
    for index, row in scotus_df.iterrows():
        # Extract case ID and argument date
        case_id = row['Case ID'].strip()
        argument_date = row['Argument Date']

        # Parse the argument date and format it for directory naming
        parsed_date = datetime.strptime(argument_date, "%m/%d/%y")

        # Skip downloading if the argument date is in the future
        if parsed_date > current_date:
            print(f"Skipping {case_id}: Argument date is in the future.")
            continue              
        
        year = parsed_date.strftime("%Y")
        month = parsed_date.strftime("%B").upper()        

        # Full path for the case directory including year, month, and case ID
        case_dir = os.path.join(f'./{year}/{month}/{case_id}/audio/')

        # Ensure the directory exists
        os.makedirs(case_dir, exist_ok=True)

        # Full path for the audio file
        audio_file_path = os.path.join(case_dir, f"{case_id}.mp3")

        # Generate the curl command
        curl_command = f"curl https://www.supremecourt.gov/media/audio/mp3files/{case_id}.mp3 --output {audio_file_path}"

        # Check if the file already exists
        if not os.path.exists(audio_file_path):
            # File doesn't exist, execute the curl command
            subprocess.run(curl_command, shell=True, check=True)
        else:
            # File already exists, skip downloading
            print(f'Skipping download, file already exists: {audio_file_path}')

# Call the function to start the download process
download_scotus_case_audio()
